crates/oxc_semantic/src/symbol/mangler.rs

```diff
@@ -0,0 +1,156 @@
+use std::cell::RefCell;
+
+use compact_str::CompactString;
+
+use super::Symbol;
+
+/// # Name Mangler / Symbol Minification
+///
+/// See:
+///   * [esbuild](https://github.com/evanw/esbuild/blob/main/docs/architecture.md#symbol-minification)
+///
+/// This algorithm is targeted for better gzip compression.
+///
+/// Visually, a slot is the index position for binding identifiers:
+///
+/// ```javascript
+/// function x(slot0, slot1) {
+///     function y(slot2, slot3) {
+///         slot0 = 1;
+///     }
+/// }
+/// function z(slot0, slot1, slot3, slot4) {
+///     slot0 = 1;
+/// }
+/// ```
+///
+/// Occurrences of slots and their corresponding newly assigned short identifiers (mangled names) are:
+/// - slot0: 4 - a
+/// - slot1: 2 - b
+/// - slot3: 2 - c
+/// - slot2: 1 - d
+/// - slot4: 1 - e
+///
+/// After swapping out the mangled names, the functions become:
+///
+/// ```javascript
+/// function x(a, b) {
+///     function y(d, c) {
+///         a = 1;
+///     }
+/// }
+/// function z(a, b, c, e) {
+///     a = 1;
+/// }
+/// ```
+#[derive(Debug, Default)]
+pub struct Mangler {
+    /// The current slot used by semantic builder
+    current_slot: Slot,
+
+    /// The maximum slot of all scopes.
+    max_slot: Slot,
+
+    /// Mangled names, indexed by slot, length by `max_slot`.
+    mangled_names: RefCell<Vec<CompactString>>,
+}
+
+/// A slot is the occurrence index of a binding identifier inside a scope.
+#[derive(Debug, Default, Clone, Copy, Eq, PartialEq, Ord, PartialOrd)]
+pub struct Slot(usize);
+
+impl Slot {
+    #[must_use]
+    #[inline]
+    pub fn index(self) -> usize {
+        self.0 - 1
```

Comment:
What if a `Slot` is constructed as `Slot(0)`? 🤔

Reply:
Let it crash ;-)